File: 算法/300. 最长递增子序列/LengthOfLIS.go
package _00__最长递增子序列

func lengthOfLIS(nums []int) int {
	// f(x)为当前位置最长递增子序列
	maxLength := 0
	dp := make([]int, len(nums))
	for j := 0; j < len(nums); j++ {
		nowMax := 1
		for i := 0; i < j; i++ {
			if nums[j] <= nums[i] {
				continue
			}
			value := dp[i] + 1
			nowMax = max(nowMax, value)
		}
		dp[j] = nowMax
		maxLength = max(maxLength, nowMax)
	}
	return maxLength
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 算法/300. 最长递增子序列/LengthOfLIS.java
public class LengthOfLIS {

    class Solution {
        public int lengthOfLIS(int[] nums) {
            int result = 0;
            int[] dp = new int[nums.length];
            for (int i = 0; i < nums.length; i++) {
                int max = 1;
                for (int j = 0; j < i; j++) {
                    if (nums[i] <= nums[j]) {
                        continue;
                    }
                    int value = dp[j] + 1;
                    max = Math.max(max, value);
                }
                dp[i] = max;
                result = Math.max(result, dp[i]);
            }
            return result;
        }
    }

}
